src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.cs

```diff
@@ -9,52 +10,270 @@ namespace Microsoft.Maui.Controls.Platform
+		{
+			var logger = _window.FindMauiContext(true)?.Services?.CreateLogger<ModalNavigationManager>();
+			SyncPlatformModalStackAsync().FireAndForget(logger, callerName);
+		}
+
+		// This code only processes a single sync action per call.
+		// It recursively calls itself until no more sync actions are left to perform.
+		//
+		// A lot can change during the process of pushing/popping a page
+		// i.e. Users might change the root page during an appearing event.
+		// So, instead of just bull dozing through the whole sync we perform one
+		// sync step then recalculate the state of affairs and then perform another
+		// until no more sync operations are left.
+		// Typically it's always a good idea to re-evaluate after any async operation has completed
+		async Task SyncPlatformModalStackAsync()
+		{
+			if (!IsModalReady || syncing)
+				return;
+
+			bool changed = false;
+
+			try
+			{
+				syncing = true;
+
+				await WindowReadyForModal();
+
+				int popTo = 0;
+
+				for (var i = 0; i < _platformModalPages.Count && i < _modalPages.Count; i++)
+				{
+					if (_platformModalPages[i] != _modalPages[i])
+					{
+						break;
+					}
+
+					popTo = i + 1;
+				}
+
+				// This means the modal stacks are already synced so we don't have to do anything
+				if (_platformModalPages.Count == _modalPages.Count && popTo == _platformModalPages.Count)
+					return;
+
+				// This ensures that appearing has fired on the final page that will be visible after 
+				// the sync has finished
+				CurrentPage?.SendAppearing();
+
+				// Pop platform modal pages until we get to the point where the xplat expectation
+				// matches the platform modals
+				if (_platformModalPages.Count > popTo && IsModalReady)
+				{
+					var page = await PopModalPlatformAsync(false);
+					page.Parent = null;
+					changed = true;
+				}
+
+				if (!changed)
 				{
-					_previousNavModel = null;
-					if (_navModel == null)
-						_navModel = new NavigationModel();
+					//push any modals that need to be synced
+					var i = _platformModalPages.Count;
+					if (i < _modalPages.Count && IsModalReady)
+					{
+						await PushModalPlatformAsync(_modalPages[i], false);
+						changed = true;
+					}
+				}
+			}
+			finally
+			{
+				syncing = false;
+			}
+
+			if (changed)
+			{
+				await SyncPlatformModalStackAsync();
+			}
+		}
+
+		public async Task<Page?> PopModalAsync(bool animated)
+		{
+			Page modal = _modalPages[_modalPages.Count - 1];
+
+			if (_window.OnModalPopping(modal))
+			{
+				_window.OnPopCanceled();
+				return null;
+			}
+
+			_modalPages.Remove(modal);
+
+			if (FireLifeCycleEvents)
+			{
+				modal.SendNavigatingFrom(new NavigatingFromEventArgs());
+				modal.SendDisappearing();
+				CurrentPage?.SendAppearing();
+			}
+
+			Task popTask;
+			if (IsModalReady)
+			{
+				popTask = PopModalPlatformAsync(animated);
+			}
+			else
+			{
+				popTask = Task.CompletedTask;
+			}
```

Comment:
Ternary operator? That will make your life simpler when you move to nullability enabled. As it is now you'll have to declare popTask as a Task? 

Using a ternary operator will remove that complaint from the compiler:
```suggestion
			popTask = IsModalReady ? PopModalPlatformAsync(animated) : Task.CompletedTask;
```